ibridges/irodsconnector/data_operations.py

```diff
@@ -110,56 +145,252 @@ def irods_put(self, local_path: str, irods_path: str, resc_name: str = ''):
+
+    def download(self, irods_path: Union[str, IrodsPath], local_path: Union[str, Path],
+                 overwrite: bool = False, resc_name: str = '', options: Optional[dict] = None):
+        """Download a collection or data object to the local filesystem
+
+        Parameters
+        ----------
+        irods_path : IrodsPath
+            Absolute irods source path pointing to a collection
+        local_path : Path
+            Absolute path to the destination directory
+        overwrite : bool
+            Overwrite existing local data
+        options : dict
+            More options for the download
+        """
+        irods_path = IrodsPath(self.session, irods_path)
+        try:
+            if irods_path.collection_exists():
+                self._download_collection(irods_path, local_path, overwrite, options)
+            else:
+                self._obj_get(irods_path, local_path, overwrite, options)
+        except irods.exception.CUT_ACTION_PROCESSED_ERR:
+            raise(irods.exception.CUT_ACTION_PROCESSED_ERR('iRODS server forbids action.'))
+
+    def get_size(self, item: Union[irods.data_object.iRODSDataObject,
+                                   irods.collection.iRODSCollection]) -> int:
+        """Collect the sizes of a data object or a
+        collection.
+
+        Parameters
+        ----------
+        item : iRODSDataObject or iRODSCollection
+
+
+        Returns
+        -------
+        int
+            Total size [bytes] of the iRODS object or all iRODS objects in the collection.
+
+        """
+        if self.is_dataobject(item):
+            return item.size
+        else:
+            all_objs = self._get_data_objects(item)
+            return sum([size for _, _, size, _ in all_objs])
+
+    def _get_data_objects(self, coll: irods.collection.iRODSCollection) -> list[str, str, int, str]:
+        """Retrieve all data objects in a collection and all its subcollections.
+
+        Parameters
+        ----------
+        coll : irods.collection.iRODSCollection
+            The collection to search for all data pbjects
+
+        Returns
+        -------
+        list of all data objects
+            [(cllection path, name, size, checksum)]
+        """
+
+        # all objects in the collection
+        objs = [(obj.collection.path, obj.name, obj.size, obj.checksum)
+                for obj in coll.data_objects]
+
+        # all objects in subcollections
+        data_query = self.session.irods_session.query(kw.COLL_NAME, kw.DATA_NAME,
+                                                      DataObject.size, DataObject.checksum)
+        data_query = data_query.filter(kw.LIKE(kw.COLL_NAME, coll.path+"/%"))
+        for res in data_query.get_results():
+            path, name, size, checksum = res.values()
+            objs.append((path, name, size, checksum))
+
+        return objs
+
+    def remove(self, irods_path: Union[IrodsPath, str]):
```

Comment:
Maybe part of IrodsPath as well?

Reply:
Yes, I already created a copy there. Once the IrodsPath is fixed, we can delete it here.